docs/kwargs.rst

```diff
@@ -0,0 +1,150 @@
+For the first attempt, it is ``0``, so it can be used in expressions
+like ``if not retry: ...``.
+
+``started`` (`datetime.datetime`) is the start time of the handler,
+in case of retries & errors -- i.e. of the first attempt.
+
+``runtime`` (`datetime.timedelta`) is the duration of the handler run,
+in case of retries & errors -- i.e. since the first attempt.
+
+
+Resource-related kwargs
+=======================
+
+.. kwarg:: body
+.. kwarg:: spec
+.. kwarg:: meta
+.. kwarg:: status
+.. kwarg:: uid
+.. kwarg:: name
+.. kwarg:: namespace
+
+Body parts
+----------
+
+``body`` is the handled object's body, a read-only mapping (dict).
+
+``spec``, ``meta``, ``status`` are aliases for relevant stanzas, and are
+live-views into ``body['spec']``, ``body['metadata']``, ``body['status']``.
+
+``namespace``, ``name``, ``uid`` can be used to identify the object being
+handled, and are aliases for the respective fields in ``body['metadata']``.
+
+
+.. kwarg:: logger
+
+Logging
+-------
+
+
+``logger`` is a per-object logger, with the messages prefixed with the object's
+namespace/name.
+
+Some of the log messages are also sent as Kubernetes events according to the
+log level configuration (default is INFO, WARNINGs, ERRORs).
+
+
+.. kwarg:: patch
+
+Patching
+--------
+
+``patch`` is a mutable mapping (dict) with the object changes to be applied
+after the handler. It is actively used internally by the framework itself,
+and is shared to the handlers for convenience _(since patching happens anyway
+in the framework, why make separate API calls for patching?)_.
+
+
+.. kwarg:: memo
+
+In-memory container
+-------------------
+
+``memo`` is an in-memory container for arbitrary runtime-only keys/fields
+and values stored during the operator lifetime.
+The values are shared by all the handlers for the same object.
+
+The in-memory values are lost on operator restarts.
+If the resource is deleted and re-created with the same name,
+the in-memory values are also lost (technically, it is a new object).
+
+
+Resource-watching kwargs
+========================
+
+For the resource watching handlers, an extra kwarg is provided:
+
+
+.. kwarg:: event
+
+API event
+---------
+
+``event`` is a raw JSON-decoded message received from Kubernetes API;
+it is a dict with ``['type']`` & ``['object']`` keys.
+
+
+Resource-changing kwargs
+========================
+
+Kopf provides functionality for change detection, and triggers the handlers
+for those changes (not for every event coming from the Kubernetes API).
+Few extra kwargs are provided for these handlers, exposing the detected changes:
+
+
+.. kwarg:: cause
+.. kwarg:: reason
+
+Causation
+---------
+
+**DEPRECATED:**
+``cause`` is the processed cause of the handler as detected by the framework (create/update/delete).
+
+``reason`` is a type of change detection (creation, update, deletion, resuming).
+It is generally reflected in the handler decorator used , but can be useful for
```

Comment:
nit-est of nits: There's a whitespace in front of the comma.

Reply:
Fixed